samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java

```diff
@@ -0,0 +1,160 @@
+
+
+public abstract class BaseCouchbaseTableFunction<V> implements InitableFunction, ClosableFunction, Serializable {
+
+  // Clients
+  private final static CouchbaseBucketRegistry COUCHBASE_BUCKET_REGISTRY = new CouchbaseBucketRegistry();
+  protected transient CouchbaseEnvironment env;
+  protected transient Cluster cluster;
+  protected transient Bucket bucket;
+
+  // Function Settings
+  protected Serde<V> valueSerde = null;
+  protected Long timeout = 0L;
+  protected TimeUnit timeUnit = null;
+  protected Integer ttl = 0; // default value 0 means no ttl, data will be stored forever
+
+  // Cluster Settings
+  protected final List<String> clusterNodes;
+  protected final String bucketName;
+
+  // Environment Settings
+  protected CouchbaseEnvironmentConfigs environmentConfigs;
+
+  public BaseCouchbaseTableFunction(String bucketName, List<String> clusterNodes, Class<V> valueClass) {
+    Preconditions.checkArgument(StringUtils.isNotEmpty(bucketName), "Bucket name is not allowed to be null or empty.");
+    Preconditions.checkArgument(CollectionUtils.isNotEmpty(clusterNodes),
+        "Cluster nodes is not allowed to be null or empty.");
+    Preconditions.checkArgument(valueClass != null, "Value class is not allowed to be null.");
+    this.bucketName = bucketName;
+    this.clusterNodes = ImmutableList.copyOf(clusterNodes);
+    this.environmentConfigs = new CouchbaseEnvironmentConfigs();
+  }
+
+  @Override
+  public void init(Context context) {
+    bucket = COUCHBASE_BUCKET_REGISTRY.getBucket(bucketName, clusterNodes, environmentConfigs);
+  }
+
+  @Override
+  public void close() {
+    COUCHBASE_BUCKET_REGISTRY.closeBucket(bucketName, clusterNodes);
+  }
+
+  public boolean isRetriable(Throwable exception) {
+    while (exception != null && !(exception instanceof TemporaryFailureException)
+        && !(exception instanceof TemporaryLockFailureException)) {
+      exception = exception.getCause();
+    }
+    return exception != null;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withTimeout(long timeout, TimeUnit timeUnit) {
+    this.timeout = timeout;
+    this.timeUnit = timeUnit;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withTtl(int ttl) {
+    this.ttl = ttl;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withSerde(Serde<V> valueSerde) {
+    this.valueSerde = valueSerde;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withUsernameAndPassword(String username, String password) {
+    environmentConfigs.username = username;
+    environmentConfigs.password = password;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withSslEnabled(boolean sslEnabled) {
+    environmentConfigs.sslEnabled = sslEnabled;
+    return (T) this;
+  }
+
+  public <T extends BaseCouchbaseTableFunction<V>> T withCertAuthEnabled(boolean certAuthEnabled) {
```

Comment:
Will it be possible to combine parameters for cert based authentication into one method just like for user/pass based?

Reply:
Hm.. I don't think we should do that. `sslEnabled`, `certAuthEnabled`, `sslKeyStoreFile`, `sslKeyStorePassword`, `sslTrustStoreFile` and `sslTrustStorePassword` are somewhat related. But they are not always used together. I think keeping them separated provides flexibility for the users.

Reply:
Can we find out the combinations?

Reply:
We can combine `sslEnabled` and `certAuthEnabled`; `sslKeyStoreFile` and `sslKeyStorePassword`; `sslTrustStoreFile` and `sslTrustStorePassword`.